### ISSUE TITLE
Configuration revamp (multiple apps support)

### ISSUE THREAD

**Issue by d4vidi** [wix/Detox#2563]
tl;dr we'd like to allow for a more flexible configuration. We're considering something like this - which motivations will be explained below (todo):

```json
{
  "detox": {
    "devices": {
      "iphone.sim": {
        "type": "ios.simulator",
        "name": "iPhone 11"
      },
      "ipad.sim": {
        "type": "ios.simulator",
        "name": "iPad Pro"
      },
      "android.emu": {
        "type": "android.emulator",
        "name": "Pixel_API_28"
      },
      "android.tablet": {
        "type": "android.emulator",
        "name": "Pixel_C_API_29"
      }
    },
    "applications": {
      "app1.release": {
        "android": {
          "binaryPath": "build/e2e.android/devRelease/app1/App.apk",
          "testBinaryPath": "./android/app/build/outputs/apk/androidTest/development/debug/app-development-debug-androidTest.apk",
          "build": "cd android ; ./gradlew assembleRelease assembleAndroidTest -DtestBuildType=debug ; cd -"
        },
        "ios": {
          "binaryPath": "build/e2e.ios/devRelease/app1/App.app",
          "build": "xcodebuild -project ios/example.xcodeproj -UseNewBuildSystem=NO -scheme example -configuration Release -sdk iphonesimulator -derivedDataPath ios/build"
        }
      },
      "app1.debug": {
        "android": {
          "binaryPath": "build/e2e.android/debug/app1/WixOneApp.apk",
          "testBinaryPath": "./android/app/build/outputs/apk/androidTest/development/debug/app-development-debug-androidTest.apk",
          "build": "cd android ; ./gradlew assembleDebug assembleAndroidTest -DtestBuildType=debug ; cd -"
        },
        "ios": {
          "binaryPath": "build/e2e.ios/debug/app1/App.app",
          "build": "xcodebuild -project ios/example.xcodeproj -UseNewBuildSystem=NO -scheme example -configuration Debug -sdk iphonesimulator -derivedDataPath ios/build"
        }
      },
      "app2.release": {
        "android": {
          "binaryPath": "build/e2e.android/devRelease/app2/App.apk",
          "testBinaryPath": "./android/app/build/outputs/apk/androidTest/development/debug/app-development-debug-androidTest.apk",
          "build": "cd android ; ./gradlew assembleRelease assembleAndroidTest -DtestBuildType=debug ; cd -"
        },
        "ios": {
          "binaryPath": "build/e2e.ios/devRelease/app2/App.app",
          "build": "xcodebuild -project ios/example.xcodeproj -UseNewBuildSystem=NO -scheme example -configuration Release -sdk iphonesimulator -derivedDataPath ios/build"
        }
      }
    },
    "configurations": {
      "android.emu.release": {
        "device": "android.emu",
        "applications": {
          "app1": "app.release",
          "app2": "app2.release"
        }
      },
      "android.tablet.release": {
        "device": "android.tablet",
        "applications": {
          "app1": "app.release",
          "app2": "app2.release"
        }
      }
    }
  }
}
```
> namely, have _devices_ and _application_ variants preliminarily defined as standalone entities, and then have the concrete configurations finalized through a simple mix-and-match lego game.

The reasons for this change are:
1. In a sense, this approach is more intuitive, and reduces the need for users to repeat themselves in their composed configuration.
2. We've been requested to make the concept of an "application" an entity managed and understood by Detox. In particular, it would eventually allow for a more flexible version of `device.installApp()` -- that accepts an app name as an argument (i.e. `device.installApp('app1')`). With that, we would effectively allow for the installation of different apps in different test suites (note: still not more than 1 app per suite...)

*Important: Must be backwards compatible, at least until the next breaking release.*

#### What would a straightforward configuration look like?

Even with just a single app, it's likely to have - in the very least, different configurations for release / debug / staging, etc. So I think this holds as-is.

#### What would an external device look like?
Under this schema, this is just another devices in the devices list -- with a path instead of a type.

#### What would a manual configuration look like?

@noomorph wdyt? `ios.none` and `ios.manual`

---

todo: see if this fits https://docs.google.com/document/d/1z2ENgzqNRHxxUfh81Illhf9Thjjgt3tGurPFg77bmtg/edit?usp=sharing

**Comment by LeoNatan** [comment]
The build command should be extended to support building all apps of a config/target, or specific apps.

`detox build --configuration android.emu.release` should build both `app.release` and `app2.release`, while a `detox build --app app1.release` should only build `app1.release`.

`applications` should probably be shortened to `apps`, to keep with the terminology used elsewhere in the project.

**Comment by d4vidi** [comment]
@LeoNatan good point, noted!

**Comment by LeoNatan** [comment]
Multiple app support is far off that a “phase 2” is a very vague concept. There are at least two major iterations of large rewrites that need to happen (API rewrite, XCUITest), I don’t see why this small feature cannot be squeezed in there.

**Comment by d4vidi** [comment]
Multiple "concurrent" apps != multiple apps config
By phase 2, I was only referring to the `build --apps` construct.
Multiple apps is a different matter. This config is more urgently required so as to allow for an easy API to test *different* apps in *different* tests.

**Comment by LeoNatan** [comment]
Implementing that before XCUITest on iOS doesn't make sense, it would require rewriting the app management stuff, just to throw that away when XCUITest is implemented and that starts taking care of app launch.

**Comment by d4vidi** [comment]
implementing multiple apps config for *different* tests requires XCUITest? Let's take it "offline"

**Comment by LeoNatan** [comment]
We can discuss here.

It depends on the implementation. Currently, app process management on iOS is handled by JS, including launch, termination, background, foreground tracking etc. This will all move to the XCUITest tester process once implemented. So I think any change in that area is a wasted effort.

**Comment by noomorph** [comment]
Well, XCUITest changes will not bring changes to JS+Android integration directly anyway, so we cannot know in advance how this is going to look like.

**Comment by LeoNatan** [comment]
It will have far-reaching changes, because iOS will require extensive modifications. I think it's better to wait. Why is this urgent anyway? In the meantime, people can just create different configs.

**Comment by d4vidi** [comment]
@LeoNatan 'not sure I follow. ATM we _can_ already provide explicit app-paths to `device.launchApp()`. This is more of a convenience just for aliasing the app's name to its binaries. I expect very little JS-side changes... The driver level shouldn't be affected AFAIU.

**Comment by d4vidi** [comment]
cc @kochavi-daniel

**Comment by LeoNatan** [comment]
Was this not implemented? @noomorph

**Comment by noomorph** [comment]
It is near the merge phase - finishing the docs...

**Comment by d4vidi** [comment]
Resolved through #2587
